math-delimiters.el

```diff
@@ -85,12 +100,88 @@
+(defcustom math-delimiters-include-characters nil
+  "Characters to include in display math mode.
+If any of them are right behind an inline math equation, slurp
+them into the display math environment and barf them back out
+when converting to inline math."
+  :group 'math-delimiters
+  :type '(list character))
+
+(defcustom math-delimiters-compressed-display-math t
+  "Whether display math should be compressed.
+A value of `t' means that, when transforming from inline to
+display math, no additional line breaks are inserted.  For
+example, we go from
+
+    text-before \\(1 + 1\\) text-after
+
+to
+
+    text-before \\=\\[1 + 1\\] text-after.
+
+A value of `nil', however, would switch between the above inline
+math version and
+
+    text-before
+    \\=\\[
+      1 + 1
+    \\]
+    text-after."
+  :group 'math-delimiters
+  :type 'boolean)
+
+(defun math-delimiters--slurp-or-barf-characters (from-close)
+  "Slurp or barf characters.
+The list of values to consider is given by
+`math-delimiters-include-characters'.  Thus, include these
+characters in display math should they come right after an inline
+math statement, and exclude them in the other direction."
+  (cl-flet ((on-char-p (pos)
+                       (cl-loop for punct in math-delimiters-include-characters
+                                do (when (= punct pos)
+                                     (cl-return t)))))
+    (if (equal from-close (cdr math-delimiters-inline)) ; from inline math
+        (when (on-char-p (char-after))
+          (forward-char))
+      (let ((orig-pos (point)))         ; from display math
+        (skip-chars-backward " \t\n" (point-min))
+        (cond ((on-char-p (char-before))
+               (backward-char))
+              (math-delimiters-compressed-display-math
+               (goto-char orig-pos)))))))
+
+(defun math-delimiters--toggle-line-breaks (open close)
+  (let* ((to-display (equal (car math-delimiters-display) open))
+         (toggle-closing `(if ,to-display
+                              (progn
+                                (TeX-newline)
+                                (search-backward ,close)
+                                (TeX-newline))
+                            (join-line -1)
+                            (join-line -1)
+                            (backward-char (1+ (length ,close)))))
+         (toggle-open `(if ,to-display
+                           (progn (TeX-newline)
+                                  (forward-char (length ,open))
+                                  (TeX-newline))
+                         (join-line)
+                         (join-line -1)
+                         ;; If necessary, like when the inline delimiters are
+                         ;; Dollars, fix `join-line's whitespace insertion.
+                         (when (equal (char-after) ?\s)
+                           (delete-char 1)))))
+    (eval toggle-closing)
+    (search-backward open)
+    (eval toggle-open))
+  (search-forward close))
```

Comment:
This is a bit spaghetti, but there was no good way I could see to make this more generic and not differentiate betwen inline -> display and display -> inline here

Reply:
We definitely don't need `eval` for this, I'll rewrite it a little differently.

Reply:
Oh, we can't use `TeX-newline`, that's an AUCTeX function. I've been careful to have the package also useable with the built-in tex modes.

Reply:
I fixed this part to use local functions instead of `eval` and to use the neutral `newline-and-indent`.

Reply:
Oh I saw "TeX-newline is [...] defined in tex.el" and didn't investigate further because surely auctex wouldn't use a file called that when there is a built in tex mode; guess I was wrong :)  Thanks for fixing these issues